Akka Http Server needs to handle bad paths with status 400 and not 500

If I create my custom `ErrorHandler extends DefaultHttpErrorHandler` , as described here https://www.playframework.com/documentation/2.6.x/ScalaErrorHandling#extending-the-default-error-handler
I have no way to change the response payload for errors like 
```
play ! @743ica805 - Internal server error, for (POST) [/action?a=X2-10GB-LRM=] ->
play
play play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 16): a=X2-10GB-LRM=
play                ^]]
play 	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:257)
play 	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:182)
play 	at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:285)
play 	at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:49)
play 	at play.core.server.Server$class.getHandlerFor(Server.scala:79)
play Caused by: akka.http.scaladsl.model.IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 16): sku=X2-10GB-LRM=
play                ^
play 	at akka.http.scaladsl.model.IllegalUriException$.apply(ErrorInfo.scala:40)
play 	at akka.http.scaladsl.model.Uri$.fail(Uri.scala:752)
play 	at akka.http.impl.model.parser.UriParser.fail(UriParser.scala:80)
play 	at akka.http.impl.model.parser.UriParser.parseQuery(UriParser.scala:69)
play 	at akka.http.scaladsl.model.Uri.query(Uri.scala:38)
```
which by the way should not be a 500, but 400 - it's a user mistyping a URL.


If I am getting it right, the reason is that Server.scala is statically coupled with DefaultHttpErrorHandler here:
https://github.com/playframework/playframework/blob/master/framework/src/play-server/src/main/scala/play/core/server/Server.scala#L53

A complete stack trace for the error:

```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 14): a=X2-10GB-LRM=
             ^]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:257)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:182)
	at play.api.http.DefaultHttpErrorHandler$.onServerError(HttpErrorHandler.scala:285)
	at play.core.server.Server$class.logExceptionAndGetResult$1(Server.scala:53)
	at play.core.server.Server$class.getHandlerFor(Server.scala:83)
	at play.core.server.AkkaHttpServer.getHandlerFor(AkkaHttpServer.scala:42)
	at play.core.server.AkkaHttpServer.getHandler(AkkaHttpServer.scala:191)
	at play.core.server.AkkaHttpServer.play$core$server$AkkaHttpServer$$handleRequest(AkkaHttpServer.scala:171)
	at play.core.server.AkkaHttpServer$$anonfun$6.apply(AkkaHttpServer.scala:93)
	at play.core.server.AkkaHttpServer$$anonfun$6.apply(AkkaHttpServer.scala:93)
Caused by: akka.http.scaladsl.model.IllegalUriException: Illegal query: Invalid input '=', expected '+', query-char, 'EOI', '&' or pct-encoded (line 1, column 14): a=X2-10GB-LRM=
             ^
	at akka.http.scaladsl.model.IllegalUriException$.apply(ErrorInfo.scala:40)
	at akka.http.scaladsl.model.Uri$.fail(Uri.scala:751)
	at akka.http.impl.model.parser.UriParser.fail(UriParser.scala:80)
	at akka.http.impl.model.parser.UriParser.parseQuery(UriParser.scala:69)
	at akka.http.scaladsl.model.Uri.query(Uri.scala:37)
	at play.core.server.akkahttp.AkkaModelConversion$$anon$2.queryMap$lzycompute(AkkaModelConversion.scala:104)
	at play.core.server.akkahttp.AkkaModelConversion$$anon$2.queryMap(AkkaModelConversion.scala:104)
	at play.api.mvc.RequestHeader$class.queryString(RequestHeader.scala:95)
	at play.api.mvc.RequestHeaderImpl.queryString(RequestHeader.scala:367)
	at play.core.routing.Route$$anon$1$$anonfun$unapply$1.apply(GeneratedRouter.scala:30)
```

--- a/framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala
+++ b/framework/src/play-akka-http-server/src/main/scala/play/core/server/akkahttp/AkkaModelConversion.scala
@@ -22,6 +22,7 @@ import play.mvc.Http.HeaderNames
 
 import scala.collection.immutable
 import scala.concurrent.Future
+import scala.util.control.NonFatal
 
 /**
  * Conversions between Akka's and Play's HTTP model objects.
@@ -99,9 +100,28 @@ private[server] class AkkaModelConversion(
       request.method.name,
       new RequestTarget {
         override lazy val uri: URI = new URI(headers.uri)
+
         override def uriString: String = headers.uri
-        override lazy val path: String = request.uri.path.toString
-        override lazy val queryMap: Map[String, Seq[String]] = request.uri.query().toMultiMap
+
+        override lazy val path: String = {
+          try {
+            request.uri.path.toString
+          } catch {
+            case NonFatal(e) =>
+              logger.warn("Failed to parse path; returning empty string.", e)
+              ""
+          }
+        }
+
+        override lazy val queryMap: Map[String, Seq[String]] = {
+          try {
+            request.uri.query().toMultiMap
+          } catch {
+            case NonFatal(e) =>
+              logger.warn("Failed to parse query string; returning empty map.", e)
+              Map[String, Seq[String]]()
+          }
+        }
       },
       request.protocol.value,
       headers,

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/BadClientHandlingSpec.scala
@@ -6,7 +6,7 @@ package play.it.http
 import play.api._
 import play.api.http.{ DefaultHttpErrorHandler, HttpErrorHandler }
 import play.api.mvc._
-import play.api.routing.Router
+import play.api.routing._
 import play.api.test._
 import play.filters.HttpFiltersComponents
 import play.it._
@@ -25,8 +25,17 @@ trait BadClientHandlingSpec extends PlaySpecification with ServerIntegrationSpec
       val port = testServerPort
 
       val app = new BuiltInComponentsFromContext(ApplicationLoader.createContext(Environment.simple())) with HttpFiltersComponents {
-        def router = Router.from {
-          case _ => defaultActionBuilder(Results.Ok)
+        def Action = defaultActionBuilder
+        def router = {
+          import sird._
+          Router.from {
+            case sird.POST(p"/action" ? q_o"query=$query") => Action { request =>
+              Results.Ok(query.getOrElse("_"))
+            }
+            case _ => Action {
+              Results.Ok
+            }
+          }
         }
         override lazy val httpErrorHandler = errorHandler
       }.application
@@ -53,6 +62,15 @@ trait BadClientHandlingSpec extends PlaySpecification with ServerIntegrationSpec
 
       response.status must_== 400
       response.body must beLeft
+    }
+
+    "still serve requests if query string won't parse" in withServer() { port =>
+      val response = BasicHttpClient.makeRequests(port)(
+        BasicRequest("POST", "/action?foo=query=bar=", "HTTP/1.1", Map(), "")
+      )(0)
+
+      response.status must_== 200
+      response.body must beLeft("_")
     }
 
     "allow accessing the raw unparsed path from an error handler" in withServer(new HttpErrorHandler() {

